lib/experimental/modules/class-gutenberg-modules.php

```diff
@@ -0,0 +1,161 @@
+<?php
+/**
+ * Gutenberg_Modules class.
+ *
+ * Native support for ES Modules and Import Maps.
+ *
+ * @package Gutenberg
+ * @subpackage Modules
+ */
+class Gutenberg_Modules {
+	/**
+	 * An array of registered modules, keyed by module identifier.
+	 *
+	 * @var array
+	 */
+	private static $registered = array();
+
+
+	/**
+	 * An array of queued modules.
+	 *
+	 * @var string[]
+	 */
+	private static $enqueued = array();
+
+	/**
+	 * Registers the module if no module with that module identifier already
+	 * exists.
+	 *
+	 * @param string $module_identifier The identifier of the module. Should be unique. It will be used in the final import map.
+	 * @param string $src               Full URL of the module, or path of the script relative to the WordPress root directory.
+	 * @param array  $args     {
+	 *      Optional array of arguments.
+	 *
+	 *      @type string|bool $ver Optional. String specifying script version number, if it has one, it is added to the URL
+	 *                             as a query string for cache busting purposes. If version is set to false, a version
+	 *                             number is automatically added equal to current installed WordPress version. If SCRIPT_DEBUG
+	 *                             is set to true, it uses the timestamp instead.
+	 * }
+	 */
+	public static function register( $module_identifier, $src, $args = array() ) {
```

Comment:
For the basic version (no server graph), I was wondering if we could support some kind of "flags" or something to say that a script is a backend script, frontend script or both. (It's a small optimization as there's a lot of backend script that are not meant to be loaded in the frontend)

Reply:
Sure. I don't think it makes sense to have a version of the API that doesn't do that.

Reply:
Added in https://github.com/WordPress/gutenberg/pull/56092/commits/11b4e1b7adb27264b003c52e9170c0eaacff5c84.